File: lib/data/models/response/order_response.dart
import 'package:json_annotation/json_annotation.dart';
import 'product_response.dart';
part 'order_response.g.dart';

@JsonSerializable()
class OrderStatusLogResponse {
  final int id;
  final String status;
  final String? note;
  final DateTime createdAt;

  const OrderStatusLogResponse({
    required this.id,
    required this.status,
    this.note,
    required this.createdAt,
  });

  factory OrderStatusLogResponse.fromJson(Map<String, dynamic> json) =>
      _$OrderStatusLogResponseFromJson(json);
  Map<String, dynamic> toJson() => _$OrderStatusLogResponseToJson(this);
}

@JsonSerializable()
class QuotationResponse {
  final int id;
  final int customerId;
  final ProductResponse? product;
  final String weightRange;
  final String cargoType;
  final String? note;
  final String status;
  final DateTime createdAt;

  const QuotationResponse({
    required this.id,
    required this.customerId,
    this.product,
    required this.weightRange,
    required this.cargoType,
    this.note,
    required this.status,
    required this.createdAt,
  });

  factory QuotationResponse.fromJson(Map<String, dynamic> json) =>
      _$QuotationResponseFromJson(json);
  Map<String, dynamic> toJson() => _$QuotationResponseToJson(this);
}

@JsonSerializable()
class OrderResponse {
  final int id;
  final int? quotationId;
  final int? customerId;
  final String? customerName;
  final String? customerPhone;
  final int? productId;
  final double totalAmount;
  final double depositAmount;
  final String status;
  final String productionStatus;
  final String? orderCode;
  final String? productName;
  final String? note;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final DateTime? estimatedDate;
  final int? assignedStaffId;
  final String? assignedStaffName;
  final List<OrderStatusLogResponse> statusLogs;

  const OrderResponse({
    required this.id,
    this.quotationId,
    this.customerId,
    this.customerName,
    this.customerPhone,
    this.productId,
    required this.totalAmount,
    required this.depositAmount,
    required this.status,
    required this.productionStatus,
    this.orderCode,
    this.productName,
    this.note,
    this.createdAt,
    this.updatedAt,
    this.estimatedDate,
    this.assignedStaffId,
    this.assignedStaffName,
    this.statusLogs = const [],
  });

  factory OrderResponse.fromJson(Map<String, dynamic> json) =>
      _$OrderResponseFromJson(json);
  Map<String, dynamic> toJson() => _$OrderResponseToJson(this);
}

File: backend/src/main/java/com/quyenauto/order/dto/OrderResponse.java
package com.quyenauto.order.dto;

import com.quyenauto.order.entity.Order;
import com.quyenauto.order.entity.OrderStatusLog;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Data
@Builder
@AllArgsConstructor
public class OrderResponse {
    private Long id;
    private String orderCode;
    private Long quotationId;
    private Long customerId;
    private String customerName;
    private String customerPhone;
    private Long productId;
    private String productName;
    private BigDecimal totalAmount;
    private BigDecimal depositAmount;
    private String status;
    private String productionStatus;
    private String note;
    private LocalDate estimatedDate;
    private Long assignedStaffId;
    private String assignedStaffName;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private List<StatusLogDto> statusLogs;

    @Data
    @Builder
    @AllArgsConstructor
    public static class StatusLogDto {
        private Long id;
        private String status;
        private String note;
        private LocalDateTime createdAt;

        public static StatusLogDto from(OrderStatusLog log) {
            return StatusLogDto.builder()
                    .id(log.getId())
                    .status(log.getStatus())
                    .note(log.getNote())
                    .createdAt(log.getCreatedAt())
                    .build();
        }
    }

    public static OrderResponse from(Order o) {
        return OrderResponse.builder()
                .id(o.getId())
                .orderCode(o.getOrderCode())
                .quotationId(o.getQuotation() != null ? o.getQuotation().getId() : null)
                .customerId(o.getCustomer().getId())
                .customerName(o.getCustomer().getFullName())
                .customerPhone(o.getCustomer().getPhone())
                .productId(o.getProduct() != null ? o.getProduct().getId() : null)
                .productName(o.getProductName())
                .totalAmount(o.getTotalAmount())
                .depositAmount(o.getDepositAmount())
                .status(o.getStatus().name())
                .productionStatus(o.getProductionStatus())
                .note(o.getNote())
                .estimatedDate(o.getEstimatedDate())
                .assignedStaffId(o.getAssignedStaff() != null ? o.getAssignedStaff().getId() : null)
                .assignedStaffName(o.getAssignedStaff() != null ? o.getAssignedStaff().getFullName() : null)
                .createdAt(o.getCreatedAt())
                .updatedAt(o.getUpdatedAt())
                .statusLogs(o.getStatusLogs().stream()
                        .map(StatusLogDto::from)
                        .collect(Collectors.toList()))
                .build();
    }
}
